File: VolumeCalculator.cs
// VolumeCalculator.cs - Калькулятор объёма тел на C# (CLI)
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text.Json;

namespace VolumeCalculator
{
    class Shape
    {
        public string Name { get; set; }
        public Func<double[], double> Func { get; set; }
        public string[] Params { get; set; }
    }

    class HistoryEntry
    {
        public string Date { get; set; }
        public string Shape { get; set; }
        public string Params { get; set; }
        public string Result { get; set; }
        public string Units { get; set; }
    }

    class Program
    {
        private static readonly double PI = Math.PI;
        private static readonly string HISTORY_FILE = "volume_history.json";
        private static readonly Dictionary<string, Shape> shapes = new Dictionary<string, Shape>();

        static Program()
        {
            shapes["1"] = new Shape { Name = "Куб", Func = p => Math.Pow(p[0], 3), Params = new[] { "сторону" } };
            shapes["2"] = new Shape { Name = "Шар", Func = p => 4.0/3.0 * PI * Math.Pow(p[0], 3), Params = new[] { "радиус" } };
            shapes["3"] = new Shape { Name = "Цилиндр", Func = p => PI * Math.Pow(p[0], 2) * p[1], Params = new[] { "радиус", "высоту" } };
            shapes["4"] = new Shape { Name = "Конус", Func = p => 1.0/3.0 * PI * Math.Pow(p[0], 2) * p[1], Params = new[] { "радиус", "высоту" } };
            shapes["5"] = new Shape { Name = "Параллелепипед", Func = p => p[0] * p[1] * p[2], Params = new[] { "длину", "ширину", "высоту" } };
            shapes["6"] = new Shape { Name = "Пирамида", Func = p => 1.0/3.0 * p[0] * p[1], Params = new[] { "площадь основания", "высоту" } };
            shapes["7"] = new Shape { Name = "Призма", Func = p => p[0] * p[1], Params = new[] { "площадь основания", "высоту" } };
            shapes["8"] = new Shape { Name = "Тор", Func = p => 2.0 * Math.Pow(PI, 2) * p[0] * Math.Pow(p[1], 2), Params = new[] { "большой радиус", "малый радиус" } };
            shapes["9"] = new Shape { Name = "Эллипсоид", Func = p => 4.0/3.0 * PI * p[0] * p[1] * p[2], Params = new[] { "полуось a", "полуось b", "полуось c" } };
            shapes["10"] = new Shape { Name = "Усечённый конус", Func = p => 1.0/3.0 * PI * p[2] * (Math.Pow(p[0], 2) + p[0]*p[1] + Math.Pow(p[1], 2)), Params = new[] { "радиус 1", "радиус 2", "высоту" } };
        }

        static void SaveHistory(HistoryEntry entry)
        {
            List<HistoryEntry> history = LoadHistory();
            history.Add(entry);
            string json = JsonSerializer.Serialize(history, new JsonSerializerOptions { WriteIndented = true });
            File.WriteAllText(HISTORY_FILE, json);
        }

        static List<HistoryEntry> LoadHistory()
        {
            if (File.Exists(HISTORY_FILE))
            {
                try
                {
                    string json = File.ReadAllText(HISTORY_FILE);
                    return JsonSerializer.Deserialize<List<HistoryEntry>>(json) ?? new List<HistoryEntry>();
                }
                catch { }
            }
            return new List<HistoryEntry>();
        }

        static void ExportCSV(string filename)
        {
            var history = LoadHistory();
            if (!history.Any())
            {
                Console.WriteLine("История пуста.");
                return;
            }
            using (var sw = new StreamWriter(filename))
            {
                sw.WriteLine("Дата,Фигура,Параметры,Объём,Единицы");
                foreach (var e in history)
                {
                    string units = string.IsNullOrEmpty(e.Units) ? "куб. ед." : e.Units;
                    sw.WriteLine($"{e.Date.Substring(0,19)},{e.Shape},\"{e.Params}\",{e.Result},{units}");
                }
            }
            Console.WriteLine($"Экспортировано в {filename}");
        }

        static double GetDouble(string prompt)
        {
            while (true)
            {
                Console.Write(prompt);
                string input = Console.ReadLine();
                if (double.TryParse(input, out double val) && val > 0)
                    return val;
                Console.WriteLine("Введите положительное число.");
            }
        }

        static void Main()
        {
            Console.WriteLine("📦 КАЛЬКУЛЯТОР ОБЪЁМА ТЕЛ");
            var history = LoadHistory();
            string units = "куб. ед.";

            while (true)
            {
                Console.WriteLine("\nВыберите фигуру:");
                foreach (var kv in shapes)
                    Console.WriteLine($"{kv.Key}. {kv.Value.Name}");
                Console.WriteLine("h. Показать историю");
                Console.WriteLine("e. Экспорт CSV");
                Console.WriteLine("u. Сменить единицы");
                Console.WriteLine("0. Выход");

                string choice = Console.ReadLine();

                if (choice == "0") break;
                else if (choice?.ToLower() == "h")
                {
                    if (!history.Any())
                    {
                        Console.WriteLine("История пуста.");
                    }
                    else
                    {
                        Console.WriteLine("\n=== ИСТОРИЯ ===");
                        foreach (var e in history.Skip(Math.Max(0, history.Count - 10)))
                            Console.WriteLine($"{e.Date.Substring(0, 19)} | {e.Shape} | {e.Result}");
                    }
                    continue;
                }
                else if (choice?.ToLower() == "e")
                {
                    Console.Write("Имя CSV файла (по умолчанию volume_history.csv): ");
                    string filename = Console.ReadLine();
                    if (string.IsNullOrEmpty(filename)) filename = "volume_history.csv";
                    ExportCSV(filename);
                    continue;
                }
                else if (choice?.ToLower() == "u")
                {
                    Console.WriteLine("Доступные единицы: куб. ед., см³, м³, дм³, л, in³");
                    Console.Write("Выберите единицы: ");
                    string input = Console.ReadLine();
                    if (!string.IsNullOrEmpty(input)) units = input;
                    continue;
                }
                else if (shapes.ContainsKey(choice))
                {
                    var shape = shapes[choice];
                    Console.WriteLine($"\nФигура: {shape.Name}");
                    var paramsList = new List<double>();
                    foreach (var pname in shape.Params)
                        paramsList.Add(GetDouble($"Введите {pname}: "));
                    double result = shape.Func(paramsList.ToArray());
                    string resultStr = $"{result:F4} {units}";
                    Console.WriteLine($"\nОбъём {shape.Name.ToLower()}: {resultStr}");

                    Console.Write("Сохранить результат? (y/n): ");
                    if (Console.ReadLine()?.ToLower() == "y")
                    {
                        var entry = new HistoryEntry
                        {
                            Date = DateTime.Now.ToString("o"),
                            Shape = shape.Name,
                            Params = string.Join(", ", paramsList),
                            Result = resultStr,
                            Units = units
                        };
                        SaveHistory(entry);
                        history.Add(entry);
                        Console.WriteLine("✅ Сохранено!");
                    }
                }
                else
                {
                    Console.WriteLine("Неверный выбор.");
                }
            }
        }
    }
}

File: VolumeCalculator.java
// VolumeCalculator.java - Калькулятор объёма тел на Java (CLI)
import java.io.*;
import java.nio.file.*;
import java.util.*;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.function.Function;

public class VolumeCalculator {
    private static final Scanner scanner = new Scanner(System.in);
    private static final String HISTORY_FILE = "volume_history.json";
    private static final double PI = Math.PI;

    static class Shape {
        String name;
        Function<List<Double>, Double> func;
        List<String> params;

        Shape(String name, Function<List<Double>, Double> func, String... params) {
            this.name = name;
            this.func = func;
            this.params = Arrays.asList(params);
        }
    }

    static class HistoryEntry {
        String date;
        String shape;
        String params;
        String result;
        String units;
    }

    private static final Map<String, Shape> shapes = new LinkedHashMap<>();

    static {
        shapes.put("1", new Shape("Куб", p -> Math.pow(p.get(0), 3), "сторону"));
        shapes.put("2", new Shape("Шар", p -> 4.0/3.0 * PI * Math.pow(p.get(0), 3), "радиус"));
        shapes.put("3", new Shape("Цилиндр", p -> PI * Math.pow(p.get(0), 2) * p.get(1), "радиус", "высоту"));
        shapes.put("4", new Shape("Конус", p -> 1.0/3.0 * PI * Math.pow(p.get(0), 2) * p.get(1), "радиус", "высоту"));
        shapes.put("5", new Shape("Параллелепипед", p -> p.get(0) * p.get(1) * p.get(2), "длину", "ширину", "высоту"));
        shapes.put("6", new Shape("Пирамида", p -> 1.0/3.0 * p.get(0) * p.get(1), "площадь основания", "высоту"));
        shapes.put("7", new Shape("Призма", p -> p.get(0) * p.get(1), "площадь основания", "высоту"));
        shapes.put("8", new Shape("Тор", p -> 2.0 * Math.pow(PI, 2) * p.get(0) * Math.pow(p.get(1), 2), "большой радиус", "малый радиус"));
        shapes.put("9", new Shape("Эллипсоид", p -> 4.0/3.0 * PI * p.get(0) * p.get(1) * p.get(2), "полуось a", "полуось b", "полуось c"));
        shapes.put("10", new Shape("Усечённый конус", p -> 1.0/3.0 * PI * p.get(2) * (Math.pow(p.get(0), 2) + p.get(0)*p.get(1) + Math.pow(p.get(1), 2)), "радиус 1", "радиус 2", "высоту"));
    }

    public static void saveHistory(HistoryEntry entry) {
        List<HistoryEntry> history = loadHistory();
        history.add(entry);
        try (PrintWriter pw = new PrintWriter(HISTORY_FILE)) {
            pw.println("[");
            for (int i = 0; i < history.size(); i++) {
                HistoryEntry e = history.get(i);
                pw.printf("  {\"date\":\"%s\",\"shape\":\"%s\",\"params\":\"%s\",\"result\":\"%s\",\"units\":\"%s\"}%s\n",
                        e.date, e.shape, e.params, e.result, e.units, (i < history.size() - 1 ? "," : ""));
            }
            pw.println("]");
        } catch (IOException ex) {}
    }

    public static List<HistoryEntry> loadHistory() {
        List<HistoryEntry> history = new ArrayList<>();
        try {
            String json = new String(Files.readAllBytes(Paths.get(HISTORY_FILE)));
            // Упрощённо: в реальном проекте использовать Jackson
        } catch (Exception e) {}
        return history;
    }

    public static void exportCSV(String filename) {
        List<HistoryEntry> history = loadHistory();
        if (history.isEmpty()) {
            System.out.println("История пуста.");
            return;
        }
        try (PrintWriter pw = new PrintWriter(filename)) {
            pw.println("Дата,Фигура,Параметры,Объём,Единицы");
            for (HistoryEntry e : history) {
                String units = e.units == null || e.units.isEmpty() ? "куб. ед." : e.units;
                pw.printf("%s,%s,\"%s\",%s,%s\n", e.date.substring(0,19), e.shape, e.params, e.result, units);
            }
            System.out.println("Экспортировано в " + filename);
        } catch (IOException ex) {
            System.out.println("Ошибка экспорта: " + ex.getMessage());
        }
    }

    public static double getDouble(String prompt) {
        while (true) {
            System.out.print(prompt);
            try {
                double val = Double.parseDouble(scanner.nextLine().trim());
                if (val > 0) return val;
                System.out.println("Введите положительное число.");
            } catch (NumberFormatException e) {
                System.out.println("Введите число.");
            }
        }
    }

    public static void main(String[] args) {
        System.out.println("📦 КАЛЬКУЛЯТОР ОБЪЁМА ТЕЛ");
        List<HistoryEntry> history = loadHistory();
        String units = "куб. ед.";

        while (true) {
            System.out.println("\nВыберите фигуру:");
            for (Map.Entry<String, Shape> entry : shapes.entrySet()) {
                System.out.println(entry.getKey() + ". " + entry.getValue().name);
            }
            System.out.println("h. Показать историю");
            System.out.println("e. Экспорт CSV");
            System.out.println("u. Сменить единицы");
            System.out.println("0. Выход");

            String choice = scanner.nextLine().trim();

            if (choice.equals("0")) break;
            else if (choice.equalsIgnoreCase("h")) {
                if (history.isEmpty()) {
                    System.out.println("История пуста.");
                } else {
                    System.out.println("\n=== ИСТОРИЯ ===");
                    int start = Math.max(0, history.size() - 10);
                    for (int i = start; i < history.size(); i++) {
                        HistoryEntry e = history.get(i);
                        System.out.printf("%s | %s | %s\n", e.date.substring(0, 19), e.shape, e.result);
                    }
                }
                continue;
            } else if (choice.equalsIgnoreCase("e")) {
                System.out.print("Имя CSV файла (по умолчанию volume_history.csv): ");
                String filename = scanner.nextLine().trim();
                if (filename.isEmpty()) filename = "volume_history.csv";
                exportCSV(filename);
                continue;
            } else if (choice.equalsIgnoreCase("u")) {
                System.out.println("Доступные единицы: куб. ед., см³, м³, дм³, л, in³");
                System.out.print("Выберите единицы: ");
                String input = scanner.nextLine().trim();
                if (!input.isEmpty()) units = input;
                continue;
            } else if (shapes.containsKey(choice)) {
                Shape shape = shapes.get(choice);
                System.out.println("\nФигура: " + shape.name);
                List<Double> params = new ArrayList<>();
                for (String pname : shape.params) {
                    params.add(getDouble("Введите " + pname + ": "));
                }
                double result = shape.func.apply(params);
                String resultStr = String.format("%.4f %s", result, units);
                System.out.printf("\nОбъём %s: %s\n", shape.name.toLowerCase(), resultStr);

                System.out.print("Сохранить результат? (y/n): ");
                String save = scanner.nextLine().trim().toLowerCase();
                if (save.equals("y")) {
                    HistoryEntry entry = new HistoryEntry();
                    entry.date = LocalDateTime.now().format(DateTimeFormatter.ISO_LOCAL_DATE_TIME);
                    entry.shape = shape.name;
                    entry.params = params.toString().replace("[", "").replace("]", "");
                    entry.result = resultStr;
                    entry.units = units;
                    saveHistory(entry);
                    history.add(entry);
                    System.out.println("✅ Сохранено!");
                }
            } else {
                System.out.println("Неверный выбор.");
            }
        }
    }
}
